### PR TITLE
Added clear button to clear search query.

### PR DESCRIPTION
**Pull request by vishwesh3**
Created custom button for Android.
clearbuttonmode only works for iOS.

Fix:#628

ed99f285-de2b-42f1-b1af-3e69c1dac0b3

### CODE REVIEW

**Comment by borisyankov** [comment]
Thanks for the PR. I was thinking that we need that also.
I looked what icon the other apps use and they use the X without background.
So it is the same as we have for iOS. Should we switch to it?

**Comment by vishwesh3** [comment]
Done 👍 
How it looks now?

iOS dark theme
78f82f72-9e43-4e1d-be58-44c8a7f3011f-2

iOS light theme
558a67fb-94c8-438d-ba1c-0f252063cf2a

Android dark theme
a6dbe09f-6802-4010-83db-6f5af3df0695

Android light theme
c861a036-058c-41bf-a19e-82899c97f046

**Comment by vishwesh3** [comment]
Updated 👍 
Rebased & fixes flow text.

File: src/common/SearchInput.js [hunk 1]
```diff
@@ -14,32 +16,52 @@ const styles = StyleSheet.create({
...
+  textInput: TextInput;
+
+  handleClear = () => {
+    this.textInput.clear();
+    this.props.onChange('');
+  }
 
   render() {
-    const { onChange } = this.props;
+    const { onChange, showCancelIcon } = this.props;
```
Comment (borisyankov) [inline src/common/SearchInput.js]: Move the `showCancelIcon` to <Input /> ?
Also call it `clearButton` as this matches the iOS version naming.

File: src/common/SearchInput.js [hunk 2]
```diff
@@ -14,32 +16,52 @@ const styles = StyleSheet.create({
...
           autoCapitalize="none"
           placeholder="Search"
           returnKeyType="search"
           onChangeText={onChange}
           autoFocus
         />
+        { showCancelIcon &&
+          <IconCross
+            style={styles.icon}
+            size={20}
```
Comment (borisyankov) [inline src/common/SearchInput.js]: The button needs to be as big as possible. Try CONTROL_SIZE from `styles.js`
Touch targets that are smaller might not work.
Suggested minimum size for iOS is 44, we might be able to get away with a bit smaller size but 20 is too small.

File: src/subscriptions/SubscriptionsScreen.js [hunk 3]
```diff
@@ -53,6 +54,7 @@ class SubscriptionsScreen extends React.Component {
         title="Subscriptions"
         searchBarOnChange={this.handleFilterChange}
         searchBar
+        showCancelIcon={filter.length > 0}
```
Comment (borisyankov) [inline src/subscriptions/SubscriptionsScreen.js]: You would want to move that to <Input /> too. The logic should always be 'if input has no text, do not show'

File: src/users/UsersScreen.js [hunk 4]
```diff
@@ -25,7 +25,10 @@ class UsersScreen extends Component {
   render() {
     const { filter } = this.state;
     return (
-      <SearchScreen searchBarOnChange={this.handleFilterChange}>
+      <SearchScreen
+        searchBarOnChange={this.handleFilterChange}
+        showCancelIcon={filter.length > 0}
```
Comment (borisyankov) [inline src/users/UsersScreen.js]: Same logic here, if moved to <Input /> will not need to be specified here.

**Comment by vishwesh3** [comment]
Updated :+1:
Button with size 44 was looking too big.
screen shot 2017-06-13 at 9 19 03 am

with size 30 (though touchable area is large than icon)
screen shot 2017-06-13 at 9 20 30 am

File: src/common/Input.js [hunk 5]
```diff
@@ -1,39 +1,74 @@
...
-    textInputRef?: (component: TextInput) => void
-  };
+
+  state: {
+    input: string,
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this.state = { input: '' };
```
Comment (borisyankov) [inline src/common/Input.js]: You an just set 
```js
state = {
  input: ''
};
```
without any constructor.

File: src/common/Input.js [hunk 6]
```diff
@@ -1,39 +1,74 @@
...
+  }
 
   static defaultProps = {
     restProps: [],
     textInputRef: () => {},
   };
 
+  handleTextChange = (text: string) => {
+    this.setState({
+      input: text,
```
Comment (borisyankov) [inline src/common/Input.js]: If you use 'input' only for 'can it be cleared' logic, you can convert to 'canBeCleared: boolean' in state.

Comment (vishwesh3) [inline src/common/Input.js]: Done :+1:

File: src/common/Input.js (hunk 5 above)
Comment (vishwesh3) [inline src/common/Input.js]: I tried removing constructor, them it was throwing null error. `this.state.canBeCleared ` is required for the initial render.

**Comment by vishwesh3** [comment]
@borisyankov rebased 👍

File: src/common/Input.js (hunk 5 above)
Comment (vishwesh3) [inline src/common/Input.js]: I got messed up. Now working 👍

**Comment by borisyankov** [comment]
This eslintrc file conflicts all the time ;)

**Comment by vishwesh3** [comment]
@borisyankov ready to go in before any other conflict arises :)

**Comment by borisyankov** [comment]
It is in! 👍

**Pull request merged**